docs/design-proposals/multi-tenancy/multi-tenancy-network.md

```diff
@@ -92,40 +94,41 @@ spec:
     ports:
       - containerPort: 443
 ```
-When a pod is attached to a certain network, it needs to set its "network" field:
+When a pod is attached to a certain network, it needs to set its "network" in labels:
 
 ```yaml
 apiVersion: v1
 kind: Pod
 metadata:
   name: nginx
+  labels:
+    arktos.futurewei.com/network: vpc-1
 spec:
   containers:
   - name: nginx
     image: nginx
     ports:
       - containerPort: 443
-  network: vpc-1
 ```
 
 
-When a pod is attached to a certain network and it wants to specify subnet or IP, it needs to be specified in pod fields:
+When a pod is attached to a certain network and it wants to specify subnet or IP, it can be expressed in annotations:
 
 ```yaml
 apiVersion: v1
 kind: Pod
 metadata:
   name: nginx
+  labels:
+    arktos.futurewei.com/network: vpc-1
+  annotations:
+    arktos.futurewei.com/nics: [{"subnet": "subnet-1", "ip": "192.168.0.12"}]
```

Comment:
I don't understand the role of nics hint. How can we make sure that they provide a concrete contract? Can you give example.

Why don't we make it even more generic and allow a network provider user input annotation (instead of the label "arktos.futurewei.com/network". For example 

User Input:
```
arktos.futurewei.com/network_provider : "Mizar"
arktos.futurewei.com/network_provider_user_parameters: "{"vpc": vpc-1, "network":  net-1}" 
```

Result (after Mizar Pod's operator):

```
arktos.futurewei.com/network-readiness=true
arktos.futurewei.com/network_provider : "Mizar"
arktos.futurewei.com/network_provider_user_parameters: "{"vpc": vpc-1, "network":  net-1}" 
arktos.futurewei.com/cni_args: "{"IPv4": 10.0.0.3, "gw":  '10.0.0.1', 'Bouncers': [B1, B2, B3]}" 
```
And then you aggregate both cni_agrs and network_provider_user_parameters as args to CNI. The same model can go with services (actually compatible to what services currently do). Right? Can you add a similar example for services.

Reply:
1. for the nics annotation, it carries the pod spec.nics field in form of annotation. More detail about spec.nics can be found [vpc-network-design](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/network/NICAndVPCSupportInArktos.md). We hope to keep the essences of nics, including the ability to have multi nics.  I'll update the example which includes nic name, hopefully it is more understandable.
2. Though nics annotation able to define intent of multiple nics, Mizar can start with single nic support.
3. I am not inclined to aggregate cni-args and user-parameters, as they serves different purpose. user-parameters is user inputs, expressing user's intent of network resource; cni-args is what the controller intends to pass on to cni plugin later as _call-back_ data. Better not mix them.
4. network-provider=Mizar seems unnecessary, as pod has network label associating the proper network object which has type (mizar in this case). Such information is consumed primarily by network controller, which should have network objects in memory already, easy to look it up. However, if this is not the case (controller does not keep a lookup table of network objects), explicit putting network type and vpc id in pod could make sense for performance sake.

Reply:
For the case of service, it has network label on creation for sure. 
If its virtual ip address is to be decided by external network provider (not sure Mizar is this kind), then the network provider can decide which hint it would like to have (nics annotation seems improper) and what sematic (like always honor the hint or OK if unsatisfied). I'll put a recommendation based on the user-input-param idea in doc; however, network provider can choose its own format if it feels like that.
If its vip is decided by Arktos, Arktos will assign vip, likely based on hints (or not, if we have no resource to impl at this stage).

Reply:
doc updated with addition of arktos.futurewei.com/vip-hint annotation.

Reply:
With given info, I came up with a solution: 

```

in common.k8s.py 

    def get_pod_annotations():
        url = ("%s/api/v1/namespaces/%s/pods/%s" % (server, namespace, pod))
        response = requests.get(url, headers=headers)
        json_response = response.json()
        annotations = json_response['metadata'].get('annotations')

        return annotations
```

```

import common.k8s

def cni_add: 
    annotations = k8s.get_pod_annotations() //Kubernetes API server
    label = k8s.get_pod_label()   //Kubernetes API server

    ip = annotations['ip_address']
    subnet = annotations['subnet']
    vpc = label['arktos.futurewei.com/network']
    ep = Endpoint(name, self.obj_api, CniService.store)
    ep.set_ip(ip)
    ep.set_vpc(vpc)
    ep.set_net(subnet)
    ...

    cni_config = {
        "cniVersion": params.cni_version,
        "interfaces": [
            {
                "name": ep.veth_name,
                "mac": ep.mac,
                "sandbox": ep.netns
            }
        ],
        "ips": [
            {
                "version": "4",
                "address": "{}/{}".format(ep.ip, ep.prefix),
                "gateway": ep.gw,
                "interface": 0
            }
        ]
    }
```

However, the major concerns with this approach is that it's depeneding on k8s API, which we really want avoid. 
We are hoping to have network info dynamically inserted into the CNI network configuration that is passed to CNI plugin binaries via stdin.
This way we can minimize dependances on k8s api.

Reply:
Another approach would be: 
1. Pod operator kicks off and create a pod object. 
2. Retrieve values from annotation. 
3. CNI gets triggered, and create network interface. 
4. Pod operator updates pod spec. 
5. Arktos gets new ip from pod's spec.

Reply:
@h-w-chen If you agree of the purpose of "arktos.futurewei.com/network_provider_user_parameters" annotations for users to pass network specific configuration, how can we separate them from cni_args annotations. My understanding is that annotations remain with the object!. Or you suggest to remove the user parameters annotations after the Network Controller processes the object and marks it network_ready?

Reply:
@zasherif 
network controller doesn't have to remove the user input annotation; leave it there is fine.
cni-args annotation is for cni usage only; since network controller and cni plugin are for same network type, controller knows what kind of data the cni plugin may need. The user input annotation, is for network controller to consume. 
yes, network controller does need to mark the pod as network-ready.

Reply:
@clu2xlu CNI_ARGS env will have whatever in pod's cni-args annotation; cni-args annotation is assigned by network controller (in your case, mizar controller). What value should be put there is totally up to mizar requirements; Arktos system only pass this data to cni plugin when cni plugin is being invoked.